res/controllers/Reloop-Beatmix-2-4-scripts.js

```diff
@@ -0,0 +1,719 @@
+
+    // toggle all jog leds
+    ReloopBeatmix24.AllJogLEDsToggle(0x90 + chan, state ? OFF : ON, 2);
+
+    var timeleft = engine.getValue(group, "duration") * (1.0 - engine.getValue(
+        group, "playposition"));
+    var nextTime = (timeleft < 15 ? 200 : 400);
+
+    if (timeleft < 30) {
+        // Set timer for leds shut off
+        jogWheelTimers[group] = engine.beginTimer(nextTime,
+            "ReloopBeatmix24.jogLedFlash(\"" + group + "\", " +
+            (state ? OFF : ON) + ")", true);
+    } else { // Back in time ?
+        // shut off all jog leds
+        ReloopBeatmix24.AllJogLEDsToggle(0x90 + chan, OFF);
+        // Lit top led
+        midi.sendShortMsg(0x90 + chan, 0x30, ON);
+
+        // Set timer for next led in 2 seconds / 16 led intervals
+        jogWheelTimers[group] = engine.beginTimer(2000 / 16,
+            "ReloopBeatmix24.nextJogLedSpinnie(\"" + group +
+            "\", 0x30)",
+            true);
+    }
+};
+
+// Effects functions
+// ========================================================
+ReloopBeatmix24.FxModeLedFlash = function(step, mode) {
+    var i;
+    var ledValue = (step % 2) ? ON : OFF;
+    if (step >= 7) {
+        for (i = 1; i <= 4; i++) {
+            // engine.trigger should be sufficient, but...
+            engine.trigger("[EffectRack1_EffectUnit1]", "group_[Channel" +
+                i + "]_enable");
+            engine.trigger("[EffectRack1_EffectUnit2]", "group_[Channel" +
+                i + "]_enable");
+            // Workaround for bug #1607277 as engine.trigger doesn't work well
+            var newValue = engine.getValue("[EffectRack1_EffectUnit1]",
+                "group_[Channel" + i + "]_enable");
+            midi.sendShortMsg(0x90 + i, 0x25, newValue ? ON : OFF);
+            midi.sendShortMsg(0x90 + i, 0x25 + SHIFT, newValue ? ON : OFF);
+            newValue = engine.getValue("[EffectRack1_EffectUnit2]",
+                "group_[Channel" + i + "]_enable");
+            midi.sendShortMsg(0x90 + i, 0x24, newValue ? ON : OFF);
+            midi.sendShortMsg(0x90 + i, 0x24 + SHIFT, newValue ? ON : OFF);
+        }
+    } else {
+        for (i = 0x91; i <= 0x94; i++) {
+            midi.sendShortMsg(i, 0x26 - mode, ledValue);
+            midi.sendShortMsg(i, 0x26 + SHIFT - mode, ledValue);
+        }
+        engine.beginTimer(150, "ReloopBeatmix24.FxModeLedFlash(" + (step + 1) +
+            ", " + mode + ")", true);
+    }
+};
+
+ReloopBeatmix24.FxModeCallback = function(group, mode) {
+    FxMode = mode;
+    FxModeLongPressed[group] = true;
+    delete FxModeTimers[group];
+    // give some visual feedback (blink led 3 times)
+    for (var i = 0x91; i <= 0x94; i++) {
+        midi.sendShortMsg(i, 0x26 - mode, OFF);
+        midi.sendShortMsg(i, 0x26 + SHIFT - mode, OFF);
+    }
+    engine.beginTimer(150, "ReloopBeatmix24.FxModeLedFlash(1, " + mode + ")",
+        true);
+};
+
+ReloopBeatmix24.ActivateFx1 = function(channel, control, value, status, group) {
```

Comment:
Also, ActivateFx1/2 could be combined to one function. Use the MIDI control number instead of hardcoded values for the EffectUnit number and FxMode numer.